### PR TITLE
ADF-4 autocomplete methods

### PR DESCRIPTION
**Pull request by kevinzetterstrom**
#### Ticket
[ADF-4](https://vimean.atlassian.net/browse/ADF-4)

#### Ticket Summary
Before we can integrate autocomplete into the Android app, we need to add the necessary groundwork in the networking layer. This ticket is about creating the API call.

#### Implementation Summary
- [x] This relies upon #181 (Search suggestion models)

I created a simple retrofit service interface and wired it up to a `Search` method, `suggest`. This method takes a query, counts for the number of results to receive per type, and a callback. 

I also created a simple test to validate that null is returned when the query is empty.

#### How to Test
You can run the `SearchTest` if you'd like, but there is nothing to test aside from that, yet.

### CODE REVIEW

File: vimeo-networking/src/test/java/com/vimeo/networking/SearchTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * Tests for the Search class
+ * <p>
+ * Created by zetterstromk on 2/22/17.
+ */
+public class SearchTest {
+
+    @Test
+    public void testSuggest_emptyQuery() throws Exception {
```
Comment (anthonycr) [inline vimeo-networking/src/test/java/com/vimeo/networking/SearchTest.java]: Nice to see a test 👍

File: vimeo-networking/src/main/java/com/vimeo/networking/model/search/VideoSuggestion.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "VideoSuggestion{" +
+               "mText='" + mText + '\'' +
+               ", mMetadata=" + mMetadata +
+               '}';
+    }
```
Comment (anthonycr) [inline vimeo-networking/src/main/java/com/vimeo/networking/model/search/VideoSuggestion.java]: Great to see equals/hashcode/tostring

File: vimeo-networking/src/main/java/com/vimeo/networking/model/search/SuggestionResponse.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        return mVideoSuggestions != null ? mVideoSuggestions.equals(that.mVideoSuggestions) : that.mVideoSuggestions == null;
+
+    }
+
+    @Override
+    public int hashCode() {
+        int result = mOndemandSuggestions != null ? mOndemandSuggestions.hashCode() : 0;
+        result = 31 * result + (mVideoSuggestions != null ? mVideoSuggestions.hashCode() : 0);
+        return result;
+    }
```
Comment (anthonycr) [inline vimeo-networking/src/main/java/com/vimeo/networking/model/search/SuggestionResponse.java]: 👍 for equals/hashcode/tostring

File: vimeo-networking/src/main/java/com/vimeo/networking/Search.java [hunk 4]
```diff
@@ -231,6 +233,49 @@ public static Facet fromString(String text) {
...
+    public static Call<SuggestionResponse> suggest(@NotNull String query, int videoSuggestionCount,
+                                                   int vodSuggestionCount,
+                                                   @NotNull ModelCallback<SuggestionResponse> callback) {
+
+        if (query.isEmpty()) {
+            callback.failure(new VimeoError("Query cannot be empty!"));
+            return null;
+        }
+
+        Map<String, String> queryMap = new HashMap<>();
```
Comment (anthonycr) [inline vimeo-networking/src/main/java/com/vimeo/networking/Search.java]: Minor perf improvement: could supply a default size of 3 here to eliminate array resizing

File: vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java [hunk 5]
```diff
@@ -1292,6 +1293,24 @@ public void failure(VimeoError error) {
...
 
     /**
+     * A package private search suggestions method. See {@link Search#suggest(String, int, int, ModelCallback)} for
+     * public usage of the suggestion API.
+     *
+     * @param queryMap the query parameters
+     * @param callback the callback to be invokedn when the call finishes
+     * @return the {@link Call} object provided by Retrofit
+     */
+    @Nullable
```
Comment (anthonycr) [inline vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java]: This should be `@NotNull` as there's no way for the call to be nullable since we dereference it immediately after getting it from the service class. If the service can return null, then we should add a null-check around the dereference.

File: vimeo-networking/src/main/java/com/vimeo/networking/model/search/SuggestionResponse.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import java.io.Serializable;
+import java.util.ArrayList;
+
+/**
+ * A response sent from the API for search/autocomplete suggestions.
+ * <p>
+ * Created by zetterstromk on 2/21/17.
+ */
+@UseStag(FieldOption.SERIALIZED_NAME)
+public class SuggestionResponse implements Serializable {
```
Comment (anthonycr) [inline vimeo-networking/src/main/java/com/vimeo/networking/model/search/SuggestionResponse.java]: It's too bad that this response can't extend from `BaseResponseList` but I guess it's fundamentally a different type of data (i.e. not pageable)

File: vimeo-networking/src/main/java/com/vimeo/networking/model/search/OndemandSuggestion.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        return "OndemandSuggestion{" +
+               "mText='" + mText + '\'' +
+               ", mMetadata=" + mMetadata +
+               '}';
+    }
```
Comment (anthonycr) [inline vimeo-networking/src/main/java/com/vimeo/networking/model/search/OndemandSuggestion.java]: 👍

**Comment by kevinzetterstrom** [comment]
@anthonycr made the changes here: 35ae4be

File: vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java [hunk 8]
```diff
@@ -1282,7 +1283,7 @@ public void failure(VimeoError error) {
      * @param callback the callback to be invoked when the call finishes
      * @return the Call object provided by the Retrofit service
      */
-    @Nullable
+    @NotNull
```
Comment (anthonycr) [inline vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java]: Great catch!
